safe_control_gym/experiment.py

```diff
@@ -1,148 +1,238 @@
     
-    def run_evaluation(self, config, ctrl, env, n_episodes=10, **kwargs):
-        """Evaluate a trained controller.
-
-        Note this evaluation uses the ctrl's own control frequency.
-        Optionally, use the ctrl's own `run()` method to evaluate instead of collecting full traj data.
+    def run_evaluation(self, n_episodes=10, log_freq=None, n_steps=None, **kwargs):
+        """Evaluate a trained controller. 
+        
+        Args:
+            n_episodes (int): number of runs to execute
+            n_steps (int): the number of steps to collect in total
+            log_freq (int): the frequency with which to log information
+        
+        Returns:
+            trajs_data (defaultdict(list)): the raw data from the executed runs 
+            metrics (dict): the metrics calculated from the raw data
         """
-        ctrl.reset()
-        if config.restore:
-            ctrl.load(os.path.join(config.restore, "model_latest.pt"))
-        trajs_data = self.collect_trajs(ctrl, env, n_trajs=n_episodes)
-        results = self.compute_metrics(trajs_data)
+
+        self.reset()
+        trajs_data = self._execute_evaluations(log_freq=log_freq, n_episodes=n_episodes, n_steps=n_steps, **kwargs)
+        metrics = self.compute_metrics(trajs_data)
+        
         # terminal printouts
-        for metric_key, metric_val in results.items():
+        for metric_key, metric_val in metrics.items():
             print("{}: {:.3f}".format(colored(metric_key,"yellow"), metric_val))
         print("Evaluation done.")
-        return results
+        return trajs_data, metrics
     
-    def collect_trajs(self, ctrl, env, log_freq=None, n_trajs=None, n_steps=None, **kwargs):
-        """"""
-        if not is_wrapped(env, RecordDataWrapper):
-            env = RecordDataWrapper(env)
+    def _execute_evaluations(self, n_episodes=10, n_steps=None, log_freq=None, **kwargs):
+        """Runs the experiments and collects all the required data.
+
+        Args:     
+            n_episodes (int): number of runs to execute
+            n_steps (int): the number of steps to collect in total
+            log_freq (int): the frequency with which to log information
+        
+        Returns:
+            trajs_data (defaultdict(list)): the raw data from the executed runs
+        """
+
+        if n_episodes is None and n_steps is None:
+            raise ValueError('One of n_episodes or n_steps must be defined.')
+        elif n_episodes is not None and n_steps is not None:
+            raise ValueError('Only one of n_episodes or n_steps can be defined.')
         
         # initialize
-        sim_steps = log_freq // env.CTRL_FREQ if log_freq else 1 
-        n_step, n_traj = 0, 0
-        done = False
+        sim_steps = log_freq // self.env.CTRL_FREQ if log_freq else 1 
+        steps, trajs = 0, 0
         ctrl_data = defaultdict(list)
-        if env.INFO_IN_RESET:
-            obs, info = env.reset()
+        if self.env.INFO_IN_RESET:
```

Comment:
given the reset of each test episode is abstracted to `_evaluation_reset()`, i think it's also possible to use it here, but provide `ctrl_data=None` to indicate nothing is lingering from the past episode.  correspondingly, add logic to deal with empty `ctrl_data` in `_evaluation_reset()`

Reply:
That's smart, will do